### PR TITLE
ASSETS-26397 [OAuth] Update Asset Compute Client

### PR DESCRIPTION
**Pull request by pheenomenon**
## Description
For https://jira.corp.adobe.com/browse/ASSETS-26397 [OAuth] Update Asset Compute Client
- Add support for OAuth server 2 server Adobe Developer console integration.
- Keep changes backward compatible
- Add tests.
- used integration at https://developer.adobe.com/console/projects/274796/4566206088344970947/overview# Org "Adobe internal - skyline ProdTest 017"

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by pheenomenon** [comment]
`Adobe CI - integration_tests` is failing due to `Too long with no output (exceeded 10m0s): context deadline exceeded` something we know to show up whenever worker-creative is updated.

File: lib/oauthservertoserver.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    return integration ? integration.TYPE === 'oauthservertoserver' : false;
+}
+
+/**
+ * Validate an OAuth Server to Server integration configuration
+ * 
+ * @param {*} integration Integration configuration
+ * @returns {boolean} True if the integration is a valid OAuth Server to Server integration
+ */
+function validateOAuthServerToServerIntegration(integration) {
```
Comment (rsathish29) [inline lib/oauthservertoserver.js:34]: I see that you are checking for the presence of the following fields. How does this confirm that its oauth? What's the diff between oauth and jwt config?

- ORGID
- CLIENT_SECRETS
- CLIENT_ID
- SCOPES
- TECHNICAL_ACCOUNT_ID
- TECHNICAL_ACCOUNT_EMAIL

File: lib/oauthservertoserver.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+ * Create an access token for an OAuth Server to Server integration
+ * 
+ * @param {*} integration Integration configuration
+ * @param {string} [adobeLoginHost] IMS host to use, defaults to https://ims-na1.adobelogin.com
+ * @returns {string} access token
+ */
+async function createOAuthServerToServerAccessToken(integration, adobeLoginHost) {
+    // API: https://wiki.corp.adobe.com/display/ims/IMS+API+-+Client+Credentials+Token
+
+    const host = adobeLoginHost || "https://ims-na1.adobelogin.com";
```
Comment (rsathish29) [inline lib/oauthservertoserver.js:53]: would it be better to validate the `adobeLoginHost` ?

File: lib/oauthservertoserver.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    const clientId = integration.CLIENT_ID;
+    const scopes = integration.SCOPES.join(',');
+
+    for (const clientSecret of integration.CLIENT_SECRETS) {
+        const formData = new FormData();
+        formData.append('client_secret', clientSecret);
+        formData.append('grant_type', 'client_credentials');
+        formData.append('scope', scopes);
+
+        const response = await fetch(`${host}/ims/token/v3?client_id=${clientId}`, {
```
Comment (rsathish29) [inline lib/oauthservertoserver.js:63]: Any reason to use v3 instead of v4?

File: package.json [hunk 4]
```diff
@@ -37,10 +40,14 @@
...
-        "test": "nyc mocha --recursive --exit",
+        "test": "nyc mocha --recursive --exit ./test",
+        "e2e": "mocha --recursive --exit ./e2e",
         "posttest": "eslint . && license-checker --summary",
         "report-coverage": "nyc report --reporter=text-lcov > coverage.lcov && codecov",
         "beautify": "eslint . --fix",
         "semantic-release": "semantic-release"
+    },
+    "engines": {
+        "node": ">=18"
```
Comment (rsathish29) [inline package.json:51]: node.js.yml refers to node-version: [14.17]. Would this have any impact?

**Review by rsathish29 — COMMENTED** [review]
Thanks for the PR @pheenomenon. Left few comments for consideration.

File: .github/workflows/node.js.yml [hunk 5]
```diff
@@ -16,7 +16,7 @@ jobs:
     if: "!contains(github.event.head_commit.message, '[ci skip]')"
     strategy:
       matrix:
-        node-version: [10.16, 10.19, 12.16, 12.19, 14.17]
+        node-version: [14.17]
```
[the remark this answers is not in the document]
Comment (pheenomenon) [reply .github/workflows/node.js.yml:19]: taking previous version of nodejs out as they break tests in general. see https://github.com/adobe/asset-compute-client/actions/runs/11469342149/job/31916394436#step:5:64 for instance.

File: e2e/register.test.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+/* eslint mocha/no-mocha-arrows: "off" */
+
+'use strict';
+
+const { AssetComputeClient, getIntegrationConfiguration } = require("../index.js");
+require("dotenv").config();
+
+describe('e2e', () => {
+    it('register', async function () {
+        // set process.env.ASSET_COMPUTE_INTEGRATION_FILE_PATH as path of the OAuth
```
[the remark this answers is not in the document]
Comment (pheenomenon) [reply e2e/register.test.js:23]: the env variable is used inside `getIntegrationConfiguration()` so comment is to set it here, before running the test. e2e is not run in CI, its mostly for developers. Auth-wize both register & process follow same flow, so didn't see a need from testing perspective.

File: lib/oauthservertoserver.js (hunk 1 above)
Comment (pheenomenon) [reply lib/oauthservertoserver.js:34]: there are sample oauth/jwt files in `resources` folder

File: lib/oauthservertoserver.js (hunk 2 above)
Comment (pheenomenon) [reply lib/oauthservertoserver.js:53]: we could if we want to fail fast, but finally it client configs and wrong config will fail.

File: lib/oauthservertoserver.js (hunk 3 above)
Comment (pheenomenon) [reply lib/oauthservertoserver.js:63]: no, nothing specific. The sample code I referred used v3 so it sticked along. will check if using v4 is compelling.

File: package.json (hunk 4 above)
Comment (pheenomenon) [reply package.json:51]: good catch. this could be a left-over. will review.

**Review by rsathish29 — APPROVED** [review]
LGTM! 👍

**Pull request merged**